compiler/modules/front/codeGen.bal

```diff
@@ -716,22 +744,67 @@ function codeGenIfElseStmt(CodeGenContext cx, bir:BasicBlock startBlock, Environ
                 elseContBlock.insns.push(branch);
                 assignments.push(...elseAssignments);
             }
-            return { block: contBlock, assignments };
+            StmtNarrowing[] narrowings = ifNarrowings;
+            foreach var n in elseNarrowings {
+                broadenStmtNarrowings(narrowings, n);
+            }
+            return { block: contBlock, assignments, narrowings };
         }
     }
 }
 
-function codeGenIfElseNarrowing(CodeGenContext cx, bir:BasicBlock bb, Environment env, Narrowing narrowing, boolean condition, Position pos) returns Environment {
+// If already exists in the list, union, else append
+function broadenStmtNarrowings(StmtNarrowing[] list, StmtNarrowing annex) {
+    foreach int i in 0..<list.length() {
+        StmtNarrowing existing = list[i];
+        if existing.binding.name == annex.binding.name {
+            list[i] = {
+                basis: { or: [annex.basis, existing.basis] },
+                binding: annex.binding,
+                ifCompletesNormally: t:union(annex.ifCompletesNormally, existing.ifCompletesNormally)
+            };
+            return;
+        }
+    }
+    list.push(annex);
+}
+
+// If already exists in the list, replace, else append
+function tightenStmtNarrowings(StmtNarrowing[] list, StmtNarrowing annex) {
+    foreach int i in 0..<list.length() {
+        StmtNarrowing existing = list[i];
+        if existing.binding.name == annex.binding.name {
+            list[i] = {
+                basis: { and: [annex.basis, existing.basis] },
+                binding: annex.binding,
+                // Assuming annex.ifCompletesNormally is a subset of existing.ifCompletesNormally
+                ifCompletesNormally: annex.ifCompletesNormally
+            };
+            return;
+        }
+    }
+    list.push(annex);
+}
+
+function codeGenIfElseNarrowing(CodeGenContext cx, bir:BasicBlock bb, Environment env, ExprNarrowing exprNarrowing, boolean condition, Position pos) returns Environment {
+    StmtNarrowing narrowing = exprNarrowingStmtNarrowing(exprNarrowing, condition);
+    Binding bindings = codeGenNarrowing(cx, bb, env, narrowing.binding, narrowing.ifCompletesNormally, narrowing.basis, pos);
+    // pr-todo: find what happens to env.narrowing
+    return { bindings, assignments: env.assignments, narrowings: [narrowing] };
+}
+
+function exprNarrowingStmtNarrowing(ExprNarrowing narrowing, boolean condition) returns StmtNarrowing {
```

Comment:
This does need a preposition. The Ballerina-ish name would be stmtNarrowingFromExprNarrowing.